simple/utils/spectral_types.py

```diff
@@ -3,320 +3,317 @@
     db,
-    sources,
-    spectral_types,
-    references,
-    regimes,
-    spectral_type_error=None,
-    comments=None,
+    source: str = None,
+    *,
+    spectral_type_string: str = None,
+    spectral_type_code: float = None,
+    spectral_type_error: float = None,
+    regime: str = None,
+    photometric: bool = False,
+    comments: str = None,
+    reference: str = None,
+    raise_error: bool = True,
 ):
     """
     Script to ingest spectral types
+
     Parameters
     ----------
     db: astrodbkit2.astrodb.Database
         Database object created by astrodbkit2
-    sources: str or list[str]
-        Names of sources
-    spectral_types: str or list[strings]
-        Spectral Types of sources
-    spectral_type_error: str or list[strings], optional
-        Spectral Type Errors of sources
-    regimes: str or list[str]
-        List or string
-    comments: list[strings], optional
+
+    source: str
+        Name of source. Constrained by the Sources table
+
+    spectral_type_string: str
+        Spectral Type of source
+
+    spectral_type_error: str, optional
+        Spectral Type Error of source
+
+    regime: str
+        String. Constrained by Regimes table
+
+    comment: str, optional
         Comments
-    references: str or list[strings]
+
+    reference: str
         Reference of the Spectral Type
+
+    raise_error: bool, optional
+
     Returns
     -------
 
     None
-
     """
-
-    n_sources = len(sources)
-
-    # Convert single element input value to list
-    input_values = [
-        sources,
-        spectral_types,
-        spectral_type_error,
-        regimes,
-        comments,
-        references,
-    ]
-    for i, input_value in enumerate(input_values):
-        if input_value is None:
-            input_values[i] = [None] * n_sources
-        elif isinstance(input_value, str):
-            input_values[i] = [input_value] * n_sources
-        # Convert single element input value to list
-    (
-        sources,
-        spectral_types,
-        spectral_type_error,
-        regimes,
-        comments,
-        references,
-    ) = input_values
-
-    n_added = 0
-    n_skipped = 0
-
-    logger.info(f"Trying to add {n_sources} spectral types")
-
-    for i, source in enumerate(sources):
-        db_name = find_source_in_db(db, source)
-        # Spectral Type data is in the database
-
-        if len(db_name) != 1:
-            msg = (
-                f"No unique source match for {source} in the database "
-                f"(with SpT: {spectral_types[i]} from {references[i]})"
+    db_name = find_source_in_db(db, source)
+
+    if len(db_name) != 1:
+        msg = f"No unique source match for {source} in the database "
+        raise AstroDBError(msg)
+    else:
+        db_name = db_name[0]
+
+    source_spt_data = (
+        db.query(db.SpectralTypes).filter(db.SpectralTypes.c.source == db_name).table()
+    )
+    logger.debug(f"Pre-existing Spectral Type data: \n {source_spt_data}")
```

Comment:
The function would flow better if lines 76-79 were after line 102. Essentially, here you are defining `source_spt_data`, then you break to check for duplicates, then you go back to analyzing `source_spt_data`. Why not have the definition right before you use it? From what I can tell the duplicate check doesn't require this.